### PR TITLE
[feat]: [CDS-78361]: Addition of custom stage to ng-manager

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
This PR contains addition of custom stage classes to cdng module. Now, custom stage classes are there in both pipeline service and in ng-manager. But pipeline-service gets preference always and it executes the custom stage deployments currently. This is done to handle in-flight pipelines when we will be removing the custom stage code from pipeline-service. (In that case, they will start using the ng-manager code of custom stage).
I have removed the code related to stage execution service. That will be handled in a separate PR. 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52601)

### CODE REVIEW

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger uts

File: pipeline-service/modules/orchestration-visualization/src/test/resources/recaster/alias/alias.txt [hunk 1]
```diff
@@ -224,4 +224,8 @@ io.harness.delegate.beans.connector.azureconnector.outcome.AzureClientSecretKeyO
 io.harness.delegate.beans.connector.azureconnector.outcome.AzureConnectorOutcomeDTO
 io.harness.delegate.beans.connector.azureconnector.outcome.AzureAuthOutcomeDTO
 io.harness.delegate.beans.connector.azureconnector.outcome.AzureCredentialOutcomeDTO
-io.harness.delegate.beans.instancesync.CustomDeploymentOutcomeMetadata
\ No newline at end of file
+io.harness.delegate.beans.instancesync.CustomDeploymentOutcomeMetadata
+io.harness.steps.shellscript.HarnessFileStoreSource
+io.harness.steps.shellscript.K8sInfraDelegateConfigOutput
+io.harness.steps.shellscript.ShellScriptInlineSource
+io.harness.steps.shellscript.ShellScriptSourceWrapper
```
Comment (lovish1999) [inline pipeline-service/modules/orchestration-visualization/src/test/resources/recaster/alias/alias.txt:231]: This is for fixing unrelated failing ut.

File: pipeline-service/modules/orchestration/src/test/resources/recaster/alias/alias.txt [hunk 2]
```diff
@@ -223,4 +223,8 @@ io.harness.delegate.beans.connector.azureconnector.outcome.AzureClientSecretKeyO
 io.harness.delegate.beans.connector.azureconnector.outcome.AzureConnectorOutcomeDTO
 io.harness.delegate.beans.connector.azureconnector.outcome.AzureAuthOutcomeDTO
 io.harness.delegate.beans.connector.azureconnector.outcome.AzureCredentialOutcomeDTO
-io.harness.delegate.beans.instancesync.CustomDeploymentOutcomeMetadata
\ No newline at end of file
+io.harness.delegate.beans.instancesync.CustomDeploymentOutcomeMetadata
+io.harness.steps.shellscript.HarnessFileStoreSource
+io.harness.steps.shellscript.K8sInfraDelegateConfigOutput
+io.harness.steps.shellscript.ShellScriptInlineSource
+io.harness.steps.shellscript.ShellScriptSourceWrapper
```
Comment (lovish1999) [inline pipeline-service/modules/orchestration/src/test/resources/recaster/alias/alias.txt:230]: This is for fixing unrelated failing ut.

File: 120-ng-manager/src/main/java/io/harness/ng/NextGenModule.java [hunk 3]
```diff
@@ -443,6 +444,17 @@ public ExecutorService templateRegistrationExecutionServiceThreadPool() {
         new ThreadFactoryBuilder().setNameFormat("FreezeTemplateRegistrationService-%d").build());
   }
 
+  @Provides
+  @Singleton
+  @Named("DashboardExecutorService")
+  public ExecutorService DashboardExecutorServiceThreadPool() {
```
Comment (ivan-83) [inline 120-ng-manager/src/main/java/io/harness/ng/NextGenModule.java:450]: Minor - DashboardExecutorServiceThreadPool >> dashboardExecutorServiceThreadPool

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> Reviewed 34 of 36 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52601)* status: 39 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/CustomStageConfig.java` line 40 at r3](https://reviewable.io/reviews/harness/harness-core/52601#-Ne2SVhODuzqf0QJnmDm:-Ne2SVhODuzqf0QJnmDn:bc5q6os) ([raw file](https://github.com/harness/harness-core/blob/5e5faa76ce92e56d4b9e877d4fdd1b99c40e70c6/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/CustomStageConfig.java#L40)):*
> ```Java
> @JsonTypeName(CUSTOM)
> @TypeAlias("CustomStageConfig")
> public class CustomStageConfig implements StageInfoConfig {
> ```

We are going to add support for `EnvironmentYamlV2 environment` in the next ticket, right?
___
*[`125-cd-nextgen/src/main/java/io/harness/serializer/morphia/NGMorphiaRegistrar.java` line 68 at r3](https://reviewable.io/reviews/harness/harness-core/52601#-Ne2VK0I2BzsAgHGpfZ1:-Ne2VK0I2BzsAgHGpfZ2:b-eyun2a) ([raw file](https://github.com/harness/harness-core/blob/5e5faa76ce92e56d4b9e877d4fdd1b99c40e70c6/125-cd-nextgen/src/main/java/io/harness/serializer/morphia/NGMorphiaRegistrar.java#L68)):*
> ```Java
>     set.add(CDLicenseReportAccounts.class);
>     set.add(AwsCdkConfig.class);
>     set.add(StageExecutionEntity.class);
> ```

stageExecutionEntity collection is in pms-harness DB. Can ng-manager connect to pms DB? We need to discuss StageExecutionEntity changes.

**Comment by lovish1999** [comment]
trigger uts

**Review by ivan-83 — DISMISSED** [review]
LGTM, just create the new ticket for moving stage execution entity logic to ng-manager CustomStageStep.

> Reviewed 21 of 21 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52601#-:-Ne3pWDK4f3DWd9SS436:b-ivpv47)* status: 39 file(s) to review. 1 of 2 approvals obtained.

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> Reviewed 18 of 18 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52601)* status: 21 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/beans/CustomStageSpecParams.java` line 23 at r5](https://reviewable.io/reviews/harness/harness-core/52601#-Ne4P-JOAMg39zG3TIXE:-Ne4P-JP0z09d1fvHdYg:bykkyjf) ([raw file](https://github.com/harness/harness-core/blob/c6997246529b73a75c36bf17c8944edefe626b6d/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/beans/CustomStageSpecParams.java#L23)):*
> ```Java
> @Builder
> @TypeAlias("CustomStageSpecParams")
> @RecasterAlias("io.harness.cdng.pipeline.beans.CustomStageSpecParams")
> ```

@archit-harness When we update the alias here and remove the custom stage from the pipeline service in the next release, will the custom stage rerun deployment work for already created/saved pipelines?

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by lovish1999** [comment]
trigger uts

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52601#-:-Ne780lt3t4XhKDe1v3t:bnfp4nl)* status: 21 file(s) to review.

**Comment by lovish1999** [comment]
trigger review

**Comment by lovish1999** [comment]
trigger messagecheck

**Comment by lovish1999** [comment]
trigger messagecheck

**Pull request merged**